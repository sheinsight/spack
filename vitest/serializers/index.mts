/**
 * 统一的快照序列化器管理
 * 按优先级组合各种专用序列化器
 */

import type { SnapshotSerializer } from 'vitest';
import { pathNormalizerSerializer } from './path-normalizer.mts';
import { duplicateDependencySerializer } from './duplicate-dependency.mts';
import { errorsArraySerializer } from './errors-array.mts';
import { generalArraySerializer } from './general-array.mts';

/**
 * 序列化器优先级列表
 * 越前面的优先级越高
 */
const serializers: SnapshotSerializer[] = [
  pathNormalizerSerializer,  // 路径标准化应该最优先
  duplicateDependencySerializer,
  errorsArraySerializer,
  generalArraySerializer,
];

/**
 * 统一快照序列化器
 * 按优先级匹配并应用相应的序列化逻辑
 */
const unifiedSerializer: SnapshotSerializer = {
  test: (val: unknown): boolean => {
    // 检查是否有任何序列化器可以处理这个值
    return serializers.some((serializer) => serializer.test(val));
  },

  serialize: (val: any) => {
    // 按优先级应用序列化逻辑
    for (const serializer of serializers) {
      if (serializer.test(val)) {
        return (serializer as SnapshotSerializer).serialize(val);
      }
    }

    // 默认情况，不应该到达这里
    return JSON.stringify(val, null, 2);
  },
};

export default unifiedSerializer;
