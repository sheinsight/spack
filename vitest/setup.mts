process.env.NODE_ENV = 'test';
process.env.API_URL = 'http://localhost:3000';
process.env.RSPACK_BINDING = `${process.cwd()}/crates/binding`;
