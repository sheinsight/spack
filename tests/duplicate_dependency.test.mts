import { test, expect } from 'vitest';
import { experiments } from '@rspack/core';
import * as binding from '@shined/spack-binding';
import { runCompiler } from './test_case.mts';

binding.registerDuplicateDependencyPlugin();
const DuplicateDependencyPlugin = experiments.createNativePlugin(
  binding.CustomPluginNames.DuplicateDependencyPlugin,
  (opt: binding.RawDuplicateDependencyPluginOpts) => ({ ...opt })
);

test('should report errors when npm alias imports have incorrect case sensitivity', async () => {
  const { promise, resolve } = Promise.withResolvers<binding.JsDuplicateDependencyPluginResp>();

  const plugin = new DuplicateDependencyPlugin({
    onDetected: (response) => resolve(response),
  });

  await runCompiler({
    fixture: 'duplicate_dependency',
    plugins: [plugin],
  });

  const response = await promise;

  expect(response.groups).toMatchSnapshot();
  for (const group of response.groups) {
    for (const lib of group.libs) {
      expect(lib.name).toContain('warning');
    }
  }
});
