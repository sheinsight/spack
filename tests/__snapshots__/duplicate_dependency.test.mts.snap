// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`should report errors when npm alias imports have incorrect case sensitivity 1`] = `
[
  {
    "name": "warning",
    "libs": [
      {
        "file": "./tests/fixtures/duplicate_dependency/node_modules/.pnpm/warning@2.1.0/node_modules/warning/package.json",
        "name": "warning",
        "version": "2.1.0"
      },
      {
        "file": "./tests/fixtures/duplicate_dependency/node_modules/.pnpm/warning@4.0.3/node_modules/warning/package.json",
        "name": "warning",
        "version": "4.0.3"
      }
    ]
  }
]
`;
